[Clover] Module not found: Error: Can't resolve '@angular/platform-server/init' ...

# 🐞 Bug report

#### What modules are related to this issue?

- [ ] builders
- [x] common
- [ ] express-engine

### Is this a regression?

Since I updated to Universal/Angular 15 next, the probleme appeared, it worked properly on last version of Universal/Angular 14.

### Description

I am trying to build the SSR server using Clover Engine, then I'm getting an error.

## 🔬 Minimal Reproduction

1. Clone https://github.com/Yberion/Angular-SSR-Clover-Engine
2. `npm run build:server:fastify`

## 🔥 Exception or Error

<pre><code>Error: Module not found: Error: Can't resolve '@angular/platform-server/init' in 'D:\GitHub\Angular-SSR-Clover-Engine'
</code></pre>

## 🌍 Your Environment

<pre><code>
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 15.0.0-rc.2
Node: 18.12.1
Package Manager: npm 8.19.2
OS: win32 x64

Angular: 15.0.0-rc.2
... cli, common, compiler, compiler-cli, core, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1500.0-rc.2
@angular-devkit/build-angular   15.0.0-rc.2
@angular-devkit/core            15.0.0-rc.2
@angular-devkit/schematics      15.0.0-rc.2
@nguniversal/common             15.0.0-next.1
@schematics/angular             15.0.0-rc.2
rxjs                            7.5.7
typescript                      4.8.4
</code></pre>

--- a/packages/angular_devkit/build_angular/src/builders/server/index.ts
+++ b/packages/angular_devkit/build_angular/src/builders/server/index.ts
@@ -24,6 +24,7 @@ import {
   generateI18nBrowserWebpackConfigFromContext,
 } from '../../utils/webpack-browser-config';
 import { getCommonConfig, getStylesConfig } from '../../webpack/configs';
+import { isPlatformServerInstalled } from '../../webpack/utils/helpers';
 import { webpackStatsLogger } from '../../webpack/utils/stats';
 import { Schema as ServerBuilderOptions } from './schema';
 
@@ -175,23 +176,21 @@ async function initialize(
 function getPlatformServerExportsConfig(wco: BrowserWebpackConfigOptions): Partial<Configuration> {
   // Add `@angular/platform-server` exports.
   // This is needed so that DI tokens can be referenced and set at runtime outside of the bundle.
-  try {
-    // Only add `@angular/platform-server` exports when it is installed.
-    // In some cases this builder is used when `@angular/platform-server` is not installed.
-    // Example: when using `@nguniversal/common/clover` which does not need `@angular/platform-server`.
-    require.resolve('@angular/platform-server', { paths: [wco.root] });
-  } catch {
-    return {};
-  }
 
-  return {
-    module: {
-      rules: [
-        {
-          loader: require.resolve('./platform-server-exports-loader'),
-          include: [path.resolve(wco.root, wco.buildOptions.main)],
+  // Only add `@angular/platform-server` exports when it is installed.
+  // In some cases this builder is used when `@angular/platform-server` is not installed.
+  // Example: when using `@nguniversal/common/clover` which does not need `@angular/platform-server`.
+
+  return isPlatformServerInstalled(wco.root)
+    ? {
+        module: {
+          rules: [
+            {
+              loader: require.resolve('./platform-server-exports-loader'),
+              include: [path.resolve(wco.root, wco.buildOptions.main)],
+            },
+          ],
         },
-      ],
-    },
-  };
+      }
+    : {};
 }

--- a/packages/angular_devkit/build_angular/src/webpack/configs/common.ts
+++ b/packages/angular_devkit/build_angular/src/webpack/configs/common.ts
@@ -42,6 +42,7 @@ import {
   getOutputHashFormat,
   getStatsOptions,
   globalScriptsByBundleName,
+  isPlatformServerInstalled,
 } from '../utils/helpers';
 
 const VENDORS_TEST = /[\\/]node_modules[\\/]/;
@@ -118,7 +119,7 @@ export async function getCommonConfig(wco: WebpackConfigOptions): Promise<Config
     // Fixes Critical dependency: the request of a dependency is an expression
     extraPlugins.push(new ContextReplacementPlugin(/@?hapi|express[\\/]/));
 
-    if (Array.isArray(entryPoints['main'])) {
+    if (isPlatformServerInstalled(wco.root) && Array.isArray(entryPoints['main'])) {
       // This import must come before any imports (direct or transitive) that rely on DOM built-ins being
       // available, such as `@angular/elements`.
       entryPoints['main'].unshift('@angular/platform-server/init');

--- a/packages/angular_devkit/build_angular/src/webpack/utils/helpers.ts
+++ b/packages/angular_devkit/build_angular/src/webpack/utils/helpers.ts
@@ -8,7 +8,6 @@
 
 import type { ObjectPattern } from 'copy-webpack-plugin';
 import { createHash } from 'crypto';
-import { existsSync } from 'fs';
 import glob from 'glob';
 import * as path from 'path';
 import type { Configuration, WebpackOptionsNormalized } from 'webpack';
@@ -305,4 +304,18 @@ export function getStatsOptions(verbose = false): WebpackStatsOptions {
   return verbose
     ? { ...webpackOutputOptions, ...verboseWebpackOutputOptions }
     : webpackOutputOptions;
+}
+
+/**
+ * @param root the workspace root
+ * @returns `true` when `@angular/platform-server` is installed.
+ */
+export function isPlatformServerInstalled(root: string): boolean {
+  try {
+    require.resolve('@angular/platform-server', { paths: [root] });
+
+    return true;
+  } catch {
+    return false;
+  }
 }

